File: lib/utils/date_parser.dart
import 'package:intl/intl.dart';
import 'package:taskly/l10n/app_localizations.dart';

/// DateParser provides intelligent date parsing with support for:
/// - Relative dates: +1d (tomorrow), +1w (next week), +1m (next month)
/// - Time-based: @now (now), @10am (10 AM today), @2pm (2 PM today)
/// - Absolute dates: 2024-12-31, 2024/12/31, 12/31/2024, 31/12/2024
///
/// Returns formatted date string in 'yyyy-MM-dd HH:mm:ss' format or null if parsing fails.
class DateParser {
  /// Parse date string with intelligent format detection.
  ///
  /// Supports:
  /// - Relative dates: +1d, +2w, +3m
  /// - Time-based: @now, @10am, @2pm
  /// - Absolute dates: 2024-12-31, 2024/12/31
  ///
  /// Returns null if input is empty or parsing fails.
  static String? parse(String input) {
    final trimmed = input.trim();
    if (trimmed.isEmpty) return null;

    try {
      if (trimmed.startsWith('+')) {
        return _parseRelativeDate(trimmed);
      } else if (trimmed.startsWith('@')) {
        return _parseTime(trimmed);
      } else {
        return _parseAbsoluteDate(trimmed);
      }
    } catch (e) {
      return null;
    }
  }

  /// Parse relative date expressions like:
  /// - +1d: tomorrow
  /// - +7d: 7 days from now
  /// - +1w: next week
  /// - +2w: 2 weeks from now
  /// - +1M: next month (30 days)
  /// - +10m: 10 minutes from now
  /// - +2h: 2 hours from now
  static String? _parseRelativeDate(String input) {
    final now = DateTime.now();
    final suffix = input.substring(1);

    final match = RegExp(r'^(\d+)([mhdwM])$').firstMatch(suffix);
    if (match != null) {
      final amount = int.parse(match.group(1)!);
      final unit = match.group(2)!;

      switch (unit) {
        case 'm':
          return _formatDateTime(now.add(Duration(minutes: amount)));
        case 'h':
          return _formatDateTime(now.add(Duration(hours: amount)));
        case 'd':
          return _formatDateTime(now.add(Duration(days: amount)));
        case 'w':
          return _formatDateTime(now.add(Duration(days: amount * 7)));
        case 'M':
          return _formatDateTime(now.add(Duration(days: amount * 30)));
      }
    }

    final lowerSuffix = suffix.toLowerCase();
    if (lowerSuffix == 'd') {
      return _formatDate(now.add(const Duration(days: 1)));
    } else if (lowerSuffix == 'w') {
      return _formatDate(now.add(const Duration(days: 7)));
    } else if (lowerSuffix == 'M') {
      return _formatDate(now.add(const Duration(days: 30)));
    }

    return null;
  }

  /// Parse time-based expressions like:
  /// - @now: current time
  /// - @10am: 10 AM today
  /// - @2pm: 2 PM today
  /// - @10:30am: 10:30 AM today
  /// - @22:30: 22:30 (24-hour format)
  /// - @10am tomorrow: 10 AM tomorrow
  /// - @10am tmw: 10 AM tomorrow (shorthand)
  /// - @8pm mon: 8 PM next Monday
  ///
  /// Converts 12-hour format to 24-hour format.
  /// If time has passed today, automatically moves to next day/week.
  static String? _parseTime(String input) {
    final now = DateTime.now();
    final timeStr = input.substring(1).toLowerCase();

    if (timeStr == 'now') {
      return _formatDateTime(now);
    }

    final parts = timeStr.split(RegExp(r'\s+'));
    if (parts.isEmpty) return null;

    final timePart = parts[0];
    final dayModifier = parts.length > 1 ? parts.sublist(1).join(' ') : null;

    int hour = 0;
    int minute = 0;

    final timeMatch = RegExp(
      r'^(\d{1,2})(?::(\d{2}))?(am|pm)?$',
    ).firstMatch(timePart);
    if (timeMatch == null) return null;

    hour = int.parse(timeMatch.group(1)!);
    final minuteStr = timeMatch.group(2);
    final period = timeMatch.group(3);

    if (minuteStr != null) {
      minute = int.parse(minuteStr);
    }

    int hour24 = hour;
    if (period == 'pm' && hour < 12) {
      hour24 = hour + 12;
    } else if (period == 'am' && hour == 12) {
      hour24 = 0;
    }

    DateTime targetDate = DateTime(
      now.year,
      now.month,
      now.day,
      hour24,
      minute,
      0,
    );

    if (dayModifier != null) {
      targetDate = _applyDayModifier(targetDate, dayModifier);
    } else if (timeStr != 'now' && targetDate.isBefore(now)) {
      targetDate = targetDate.add(const Duration(days: 1));
    }

    return _formatDateTime(targetDate);
  }

  /// Apply day modifier to a date.
  /// Supports: tomorrow, tmw, mon, tue, wed, thu, fri, sat, sun
  static DateTime _applyDayModifier(DateTime date, String modifier) {
    final lowerModifier = modifier.toLowerCase().trim();

    if (lowerModifier == 'tomorrow' || lowerModifier == 'tmw') {
      return date.add(const Duration(days: 1));
    }

    final daysMap = {
      'mon': 1,
      'tue': 2,
      'wed': 3,
      'thu': 4,
      'fri': 5,
      'sat': 6,
      'sun': 7,
    };

    final targetDay = daysMap[lowerModifier];
    if (targetDay != null) {
      final currentDay = date.weekday;
      int daysToAdd = targetDay - currentDay;
      if (daysToAdd <= 0) {
        daysToAdd += 7;
      }
      return date.add(Duration(days: daysToAdd));
    }

    return date;
  }

  /// Parse absolute date strings in various formats:
  /// - yyyy-MM-dd: 2024-12-31
  /// - yyyy/MM/dd: 2024/12/31
  /// - MM/dd/yyyy: 12/31/2024
  /// - dd/MM/yyyy: 31/12/2024
  ///
  /// Tries each format in order until one succeeds.
  static String? _parseAbsoluteDate(String input) {
    final formats = ['yyyy-MM-dd', 'yyyy/MM/dd', 'MM/dd/yyyy', 'dd/MM/yyyy'];

    for (final format in formats) {
      try {
        final date = DateFormat(format).parseStrict(input);
        return _formatDate(date);
      } catch (e) {
        continue;
      }
    }

    return null;
  }

  /// Format DateTime to 'yyyy-MM-dd HH:mm:ss' string.
  static String _formatDate(DateTime date) {
    return DateFormat('yyyy-MM-dd HH:mm:ss').format(date);
  }

  /// Format DateTime to 'yyyy-MM-dd HH:mm:ss' string.
  static String _formatDateTime(DateTime date) {
    return DateFormat('yyyy-MM-dd HH:mm:ss').format(date);
  }

  /// Format date string for user-friendly display.
  ///
  /// Returns:
  /// - 'Today HH:mm' for today
  /// - 'Tomorrow HH:mm' for tomorrow
  /// - 'Yesterday HH:mm' for yesterday
  /// - 'yyyy-MM-dd HH:mm' for other dates
  ///
  /// Returns original string if parsing fails.
  static String formatDateForDisplay(String? dateStr, AppLocalizations l10n) {
    if (dateStr == null || dateStr.isEmpty) return '';
    try {
      final date = DateTime.parse(dateStr);
      final now = DateTime.now();
      final today = DateTime(now.year, now.month, now.day);
      final taskDate = DateTime(date.year, date.month, date.day);

      if (taskDate == today) {
        return '${l10n.navToday} ${DateFormat('HH:mm').format(date)}';
      } else if (taskDate == today.add(const Duration(days: 1))) {
        return '${l10n.dateTomorrow} ${DateFormat('HH:mm').format(date)}';
      } else if (taskDate == today.subtract(const Duration(days: 1))) {
        return '${l10n.dateYesterday} ${DateFormat('HH:mm').format(date)}';
      } else {
        return DateFormat('yyyy-MM-dd HH:mm').format(date);
      }
    } catch (e) {
      return dateStr;
    }
  }

  /// Format date part only for user-friendly display.
  ///
  /// Returns:
  /// - 'Today' for today
  /// - 'Tomorrow' for tomorrow
  /// - 'Yesterday' for yesterday
  /// - 'yyyy-MM-dd' for other dates
  ///
  /// Returns original string if parsing fails.
  static String formatDateOnlyForDisplay(String? dateStr, AppLocalizations l10n) {
    if (dateStr == null || dateStr.isEmpty) return '';
    try {
      final date = DateTime.parse(dateStr);
      final now = DateTime.now();
      final today = DateTime(now.year, now.month, now.day);
      final taskDate = DateTime(date.year, date.month, date.day);

      if (taskDate == today) {
        return l10n.navToday;
      } else if (taskDate == today.add(const Duration(days: 1))) {
        return l10n.dateTomorrow;
      } else if (taskDate == today.subtract(const Duration(days: 1))) {
        return l10n.dateYesterday;
      } else {
        return DateFormat('yyyy-MM-dd').format(date);
      }
    } catch (e) {
      return dateStr;
    }
  }

  /// Format time part only for display.
  ///
  /// Input format: 'HH:mm' or 'yyyy-MM-dd HH:mm:ss'
  /// Returns: 'HH:mm' format
  ///
  /// Returns empty string if parsing fails.
  static String formatTimeForDisplay(String? timeStr) {
    if (timeStr == null || timeStr.isEmpty) return '';
    try {
      // If it's a full datetime string, extract time
      if (timeStr.contains('-')) {
        final date = DateTime.parse(timeStr);
        return DateFormat('HH:mm').format(date);
      }
      // If it's already a time string, validate and return
      final parts = timeStr.split(':');
      if (parts.length >= 2) {
        final hour = int.tryParse(parts[0]);
        final minute = int.tryParse(parts[1]);
        if (hour == null || minute == null) return timeStr;
        if (hour >= 0 && hour < 24 && minute >= 0 && minute < 60) {
          return timeStr;
        }
      }
      return timeStr;
    } catch (e) {
      return '';
    }
  }

  /// Combine date and time into a full datetime string.
  ///
  /// Takes separate date (yyyy-MM-dd) and time (HH:mm) strings
  /// and combines them into 'yyyy-MM-dd HH:mm:ss' format.
  static String combineDateTime(String dateStr, String? timeStr) {
    if (timeStr == null || timeStr.isEmpty) {
      return '$dateStr 00:00:00';
    }
    try {
      final date = DateFormat('yyyy-MM-dd').parse(dateStr);
      final parts = timeStr.split(':');
      final hour = int.parse(parts[0]);
      final minute = int.parse(parts[1]);
      final combined = DateTime(date.year, date.month, date.day, hour, minute);
      return DateFormat('yyyy-MM-dd HH:mm:ss').format(combined);
    } catch (e) {
      return '$dateStr 00:00:00';
    }
  }

  /// Extract time command from task text.
  ///
  /// Returns a tuple of (taskText, timeCommand) where:
  /// - taskText: task text without time command
  /// - timeCommand: extracted time command (e.g., '+1d', '@10am', etc.)
  ///
  /// Examples:
  /// - "Buy milk @10am" → ("Buy milk", "@10am")
  /// - "Meeting +2h" → ("Meeting", "+2h")
  /// - "Go to gym @8pm mon" → ("Go to gym", "@8pm mon")
  /// - "Write report" → ("Write report", null)
  ///
  /// Time commands can be:
  /// - Relative: +10m, +2h, +1d, +1w, +6M
  /// - Exact: @now, @10am, @10:30pm, @22:30
  /// - With day modifier: @10am tomorrow, @8pm mon
  static (String, String?) extractTimeCommand(String input) {
    final trimmed = input.trim();
    if (trimmed.isEmpty) return ('', null);

    final patterns = [
      r'(?:^|\s)([+]\d+[mhdwM])(?:\s|$)',
      r'(?:^|\s)([@](?:now|\d{1,2}(?::\d{2})?(?:am|pm)?)(?:\s+(?:tomorrow|tmw|mon|tue|wed|thu|fri|sat|sun))?)\s*$',
    ];

    for (final pattern in patterns) {
      final timePattern = RegExp(pattern, caseSensitive: false);
      final match = timePattern.firstMatch(trimmed);
      if (match != null) {
        final timeCommand = match.group(1)!.trim();
        final taskText = trimmed.substring(0, match.start).trim();
        return (taskText, timeCommand);
      }
    }

    return (trimmed, null);
  }
}

File: src/Taskly/Services/DateParser.cs
using System.Globalization;
using System.Text.RegularExpressions;

namespace Taskly.Services;

/// <summary>
/// 自然语言日期/时间解析器，对应原 Flutter 版 utils/date_parser.dart。
/// 支持调度速记：
///   +10m (10 分钟), +2h (2 小时), +1d (1 天), +1w (1 周), +1M (1 月=30 天)
///   @now, @10am, @2pm, @10:30am, @22:30, @10am tomorrow/tmw, @8pm mon(tue/wed/thu/fri/sat/sun)
///   绝对日期: yyyy-MM-dd / yyyy/MM/dd / MM/dd/yyyy / dd/MM/yyyy
///   extractTimeCommand: 从文本尾部提取时间命令（如 "买牛奶 @10am" → ("买牛奶", "@10am")）
/// 返回格式：解析结果为 'yyyy-MM-dd HH:mm:ss'；纯日期为 'yyyy-MM-dd'；纯时间为 'HH:mm'。
/// 注意：m=分钟，M=月（与原版正则单位一致）。
/// </summary>
public sealed class DateParser
{
    // 相对时间：+数字+单位。单位 m=分钟 h=小时 d=天 w=周 M=月(30天)
    private static readonly Regex RelativeDateRegex = new(
        @"^(\d*)([mhdwM])$",
        RegexOptions.CultureInvariant);

    // 时间：@now / @10 / @10am / @10:30 / @10:30am / @22:30 / @22:30pm
    private static readonly Regex TimeRegex = new(
        @"^(\d{1,2})(?::(\d{2}))?(am|pm)?$",
        RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);

    // 提取尾部相对命令：+10m / +2h 等
    private static readonly Regex ExtractRelativeRegex = new(
        @"(?:^|\s)(\+\d+[mhdwM])(?:\s|$)",
        RegexOptions.CultureInvariant);

    // 提取末尾 @ 命令（含可选的修饰符后缀）
    private static readonly Regex ExtractAtRegex = new(
        @"@(?:now|\d{1,2}(?::\d{2})?(?:am|pm)?)(?:\s+(?:tomorrow|tmw|mon|tue|wed|thu|fri|sat|sun))?$",
        RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);

    /// <summary>
    /// 解析输入字符串为完整日期时间 'yyyy-MM-dd HH:mm:ss'，无法解析返回 null。
    /// </summary>
    public string? Parse(string input)
    {
        if (string.IsNullOrWhiteSpace(input))
        {
            return null;
        }

        var s = input.Trim();

        if (s.StartsWith('+'))
        {
            return ParseRelativeDate(s);
        }

        if (s.StartsWith('@'))
        {
            return ParseAtTime(s[1..]);
        }

        return ParseAbsoluteDate(s);
    }

    /// <summary>
    /// 从文本中提取尾部的时间命令，返回 (剩余文本, 时间命令原始字符串)。
    /// 时间命令为 null 表示未提取到。对应原版 extractTimeCommand。
    /// </summary>
    public (string Text, string? TimeCommand) ExtractTimeCommand(string input)
    {
        if (string.IsNullOrWhiteSpace(input))
        {
            return (input ?? string.Empty, null);
        }

        var text = input;
        string? command = null;

        // 先尝试末尾 @ 命令
        var atMatch = ExtractAtRegex.Match(text);
        if (atMatch.Success)
        {
            var atIndex = text.IndexOf('@');
            command = text[atIndex..].Trim();
            text = text[..atIndex].TrimEnd();
        }

        // 再尝试尾部相对命令
        var relMatch = ExtractRelativeRegex.Match(text);
        if (relMatch.Success)
        {
            command = relMatch.Groups[1].Value;
            // 去掉匹配到的命令片段（含可能的前导空格）
            text = text.Remove(relMatch.Index, relMatch.Length).Trim();
        }

        return (text.Trim(), command);
    }

    // ---------------- 相对时间 +数字+单位 ----------------

    private string? ParseRelativeDate(string input)
    {
        var body = input[1..]; // 去掉 +
        var match = RelativeDateRegex.Match(body);
        if (!match.Success)
        {
            return null;
        }

        // 无数字时默认 1
        var num = match.Groups[1].Value;
        var amount = string.IsNullOrEmpty(num) ? 1 : int.Parse(num, CultureInfo.InvariantCulture);
        var unit = match.Groups[2].Value;

        var now = DateTime.Now;
        var result = unit switch
        {
            "m" => now.AddMinutes(amount),
            "h" => now.AddHours(amount),
            "d" => now.AddDays(amount),
            "w" => now.AddDays(amount * 7),
            "M" => now.AddDays(amount * 30), // 月 = 30 天（与原版一致）
            _ => (DateTime?)null,
        };

        return result?.ToString("yyyy-MM-dd HH:mm:ss", CultureInfo.InvariantCulture);
    }

    // ---------------- @ 时间 ----------------

    private string? ParseAtTime(string input)
    {
        var trimmed = input.Trim();

        if (trimmed.Equals("now", StringComparison.OrdinalIgnoreCase))
        {
            return DateTime.Now.ToString("yyyy-MM-dd HH:mm:ss", CultureInfo.InvariantCulture);
        }

        // 拆分时间主体和可选的日期修饰符（tomorrow/tmw/工作日）
        var parts = trimmed.Split(new[] { ' ' }, 2, StringSplitOptions.RemoveEmptyEntries);
        var timePart = parts[0];
        var modifier = parts.Length > 1 ? parts[1] : null;

        var match = TimeRegex.Match(timePart);
        if (!match.Success)
        {
            return null;
        }

        var hour = int.Parse(match.Groups[1].Value, CultureInfo.InvariantCulture);
        var min = match.Groups[2].Success ? int.Parse(match.Groups[2].Value, CultureInfo.InvariantCulture) : 0;
        var ampm = match.Groups[3].Success ? match.Groups[3].Value.ToLowerInvariant() : null;

        if (ampm == "am")
        {
            if (hour == 12)
            {
                hour = 0;
            }
        }
        else if (ampm == "pm")
        {
            if (hour < 12)
            {
                hour += 12;
            }
        }

        if (hour is < 0 or > 23 || min is < 0 or > 59)
        {
            return null;
        }

        var today = DateTime.Now;
        var date = new DateTime(today.Year, today.Month, today.Day, hour, min, 0);

        // 应用日期修饰符
        if (!string.IsNullOrEmpty(modifier))
        {
            date = ApplyDayModifier(date, modifier!);
        }
        else if (date < DateTime.Now)
        {
            // 时间已过且无修饰符，顺延到明天（与原版一致）
            date = date.AddDays(1);
        }

        return date.ToString("yyyy-MM-dd HH:mm:ss", CultureInfo.InvariantCulture);
    }

    private DateTime ApplyDayModifier(DateTime date, string modifier)
    {
        var m = modifier.ToLowerInvariant();
        if (m is "tomorrow" or "tmw")
        {
            return date.AddDays(1);
        }

        // 工作日：下一个该工作日（若今天已是该日或已过，则下周）
        var targetDay = m switch
        {
            "sun" => DayOfWeek.Sunday,
            "mon" => DayOfWeek.Monday,
            "tue" => DayOfWeek.Tuesday,
            "wed" => DayOfWeek.Wednesday,
            "thu" => DayOfWeek.Thursday,
            "fri" => DayOfWeek.Friday,
            "sat" => DayOfWeek.Saturday,
            _ => (DayOfWeek?)null,
        };

        if (targetDay is { } td)
        {
            var diff = ((int)td - (int)date.DayOfWeek + 7) % 7;
            if (diff <= 0)
            {
                diff += 7;
            }

            return date.AddDays(diff);
        }

        return date;
    }

    // ---------------- 绝对日期 ----------------

    private string? ParseAbsoluteDate(string input)
    {
        var s = input.Trim();

        string[]? formats =
        {
            "yyyy-MM-dd", "yyyy/MM/dd", "MM/dd/yyyy", "dd/MM/yyyy",
        };

        foreach (var fmt in formats)
        {
            if (DateTime.TryParseExact(s, fmt, CultureInfo.InvariantCulture, DateTimeStyles.None, out var date))
            {
                if (date.Year < ValidationHelper.MinYear || date.Year > ValidationHelper.MaxYear)
                {
                    return null;
                }

                return date.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture);
            }
        }

        return null;
    }

    // ---------------- 显示格式化（对应原版 formatDateForDisplay 等）----------------

    /// <summary>从 'yyyy-MM-dd HH:mm:ss' 或 'yyyy-MM-dd' 解析出日期部分（yyyy-MM-dd）。</summary>
    public static string? ExtractDateOnly(string? dateTimeStr)
    {
        if (string.IsNullOrEmpty(dateTimeStr))
        {
            return null;
        }

        // 截取前 10 位 yyyy-MM-dd
        return dateTimeStr!.Length >= 10 ? dateTimeStr[..10] : dateTimeStr;
    }

    /// <summary>从 'yyyy-MM-dd HH:mm:ss' 解析出时间部分（HH:mm）。无时间返回 null。</summary>
    public static string? ExtractTimeOnly(string? dateTimeStr)
    {
        if (string.IsNullOrEmpty(dateTimeStr) || dateTimeStr!.Length < 16)
        {
            return null;
        }

        return dateTimeStr[11..16];
    }

    /// <summary>日期+时间合并为 'yyyy-MM-dd HH:mm:ss'（无时间则 00:00:00）。对应原版 combineDateTime。</summary>
    public static string CombineDateTime(string? dateStr, string? timeStr)
    {
        var date = ExtractDateOnly(dateStr) ?? DateTime.Now.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture);
        var time = !string.IsNullOrEmpty(timeStr) ? timeStr : "00:00";
        return $"{date} {time}:00";
    }

    /// <summary>仅日期显示（今天/明天/昨天/yyyy-MM-dd）。对应原版 formatDateOnlyForDisplay。</summary>
    public string FormatDateOnlyForDisplay(string? dateStr, Func<string> todayLabel,
        Func<string> tomorrowLabel, Func<string> yesterdayLabel)
    {
        var dateOnly = ExtractDateOnly(dateStr);
        if (string.IsNullOrEmpty(dateOnly))
        {
            return string.Empty;
        }

        if (DateTime.TryParseExact(dateOnly, "yyyy-MM-dd", CultureInfo.InvariantCulture,
            DateTimeStyles.None, out var date))
        {
            var today = DateTime.Today;
            if (date.Date == today)
            {
                return todayLabel();
            }

            if (date.Date == today.AddDays(1))
            {
                return tomorrowLabel();
            }

            if (date.Date == today.AddDays(-1))
            {
                return yesterdayLabel();
            }
        }

        return dateOnly!;
    }

    /// <summary>日期+时间显示（今天 HH:mm 等）。对应原版 formatDateForDisplay。</summary>
    public string FormatDateTimeForDisplay(string? dateTimeStr, Func<string> todayLabel,
        Func<string> tomorrowLabel, Func<string> yesterdayLabel)
    {
        if (string.IsNullOrEmpty(dateTimeStr))
        {
            return string.Empty;
        }

        var dateOnly = ExtractDateOnly(dateTimeStr);
        var timeOnly = ExtractTimeOnly(dateTimeStr);
        var dateLabel = FormatDateOnlyForDisplay(dateOnly, todayLabel, tomorrowLabel, yesterdayLabel);

        return string.IsNullOrEmpty(timeOnly)
            ? dateLabel
            : $"{dateLabel} {timeOnly}";
    }
}
